lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java

```diff
@@ -112,136 +194,182 @@ public boolean stop(long timeout, @NonNull TimeUnit unit) {
+            synchronized (this) {
+                if (stopLatch != null) { throw new ExecutorClosedException("Executor has been stopped"); }
+
+                pendingTasks.add(new InstrumentedTask(task, this::scheduleNext));
 
-            if (running == null) { scheduleNext(); }
+                if (needsRestart || (pendingTasks.size() <= 1)) { executeTask(null); }
+            }
         }
 
+        /**
+         * Stop the executor.
+         * If this call returns false, the executor has *not* yet stopped.
+         * It will continue to run tasks from its queue until all have completed.
+         *
+         * @param timeout time to wait for shutdown
+         * @param unit    time unit for shutdown wait
+         * @return true if all currently scheduled tasks completed before the shutdown
+         */
         @Override
         public boolean stop(long timeout, @NonNull TimeUnit unit) {
+            Preconditions.testArg(timeout, "timeout must be >= 0", x -> x >= 0);
+            Preconditions.checkArgNotNull(unit, "time unit");
+
             final CountDownLatch latch;
             synchronized (this) {
-                if (stopLatch == null) {
-                    stopLatch = new CountDownLatch((running == null) ? 0 : tasks.size() + 1);
-                }
-
+                if (stopLatch == null) { stopLatch = new CountDownLatch(1); }
+                if (pendingTasks.size() <= 0) { return true; }
                 latch = stopLatch;
             }
 
-            if (latch.getCount() <= 0) { return true; }
-
             try { return latch.await(timeout, unit); }
             catch (InterruptedException ignore) { }
 
             return false;
         }
 
+        // Called on completion of the task at the head of the pending queue.
         private void scheduleNext() {
+            final CountDownLatch latch;
             synchronized (this) {
-                if (stopLatch != null) { stopLatch.countDown(); }
+                executeTask(pendingTasks.remove());
+                latch = (pendingTasks.size() > 0) ? null : stopLatch;
+            }
 
-                final InstrumentedTask prev = running;
-                running = tasks.poll();
-                if (running == null) { return; }
+            if (latch != null) { latch.countDown(); }
+        }
 
-                try { executor.execute(running); }
-                catch (RejectedExecutionException e) {
-                    dumpExecutorState(e, prev, running, tasks);
-                    running = null;
-                }
+        @GuardedBy("this")
+        private void executeTask(@Nullable InstrumentedTask prevTask) {
+            final InstrumentedTask nextTask = pendingTasks.peek();
+            if (nextTask == null) { return; }
+
+            try {
+                executor.execute(nextTask);
+                needsRestart = false;
+            }
+            catch (RejectedExecutionException e) {
+                needsRestart = true;
```

Comment:
What if `execute` is never called again?  Won't this executor just stop forever?

Reply:
Same question here.

Reply:
This implementation attempts to restart the queue when client code submits a new task
It was the best I could come up with.  If you can think of something better that would be great.
Here is my analysis:
There are no execution resources.  Attempting to re-schedule the task is pointless and, essentially, a spin lock.  Attempting to reschedule the task after a delay might make some sense.  In the normal case, though -- a permanently stalled task --  it is also pointless.
I've thought about some mechanism that would throw the REE to the use when they attempt to submit to a stalled queue.  That is definitely appealing...